Modules/_asynciomodule.c

```diff
@@ -518,6 +578,129 @@ future_init(FutureObj *fut, PyObject *loop)
     return 0;
 }
 
+static int
+future_awaited_by_add_lock_held(asyncio_state *state, PyObject *fut, PyObject *thing)
+{
+    FutureObj *_fut = (FutureObj *)fut;
+
+    /* Most futures/task are only awaited by one entity, so we want
+       to avoid always creating a set for `fut_awaited_by`.
+    */
+    if (_fut->fut_awaited_by == NULL) {
+        assert(!_fut->fut_awaited_by_is_set);
+        Py_INCREF(thing);
+        _fut->fut_awaited_by = thing;
+        return 0;
+    }
+
+    if (_fut->fut_awaited_by_is_set) {
+        assert(PySet_Check(_fut->fut_awaited_by));
+        return PySet_Add(_fut->fut_awaited_by, thing);
+    }
+
+    PyObject *set = PySet_New(NULL);
+    if (set == NULL) {
+        return -1;
+    }
+    if (PySet_Add(set, thing)) {
+        Py_DECREF(set);
+        return -1;
+    }
+    if (PySet_Add(set, _fut->fut_awaited_by)) {
+        Py_DECREF(set);
+        return -1;
+    }
+    Py_SETREF(_fut->fut_awaited_by, set);
+    _fut->fut_awaited_by_is_set = 1;
+    return 0;
+}
+
+static int
+future_awaited_by_add(asyncio_state *state, PyObject *fut, PyObject *thing)
+{
+    if (!TaskOrFuture_Check(state, fut) || !TaskOrFuture_Check(state, thing)) {
+        // We only want to support native asyncio Futures.
+        // For further insight see the comment in the Python
+        // implementation of "future_add_to_awaited_by()".
+        return 0;
+    }
+
+    int result;
+    Py_BEGIN_CRITICAL_SECTION(fut);
+    result = future_awaited_by_add_lock_held(state, fut, thing);
+    Py_END_CRITICAL_SECTION();
+    return result;
+}
+
+static int
+future_awaited_by_discard_lock_held(asyncio_state *state, PyObject *fut, PyObject *thing)
+{
+
+    FutureObj *_fut = (FutureObj *)fut;
+
+    /* Following the semantics of 'set.discard()' here in not
+       raising an error if `thing` isn't in the `awaited_by` "set".
+    */
+    if (_fut->fut_awaited_by == NULL) {
+        return 0;
+    }
+    if (_fut->fut_awaited_by == thing) {
+        Py_CLEAR(_fut->fut_awaited_by);
+        return 0;
+    }
+    if (_fut->fut_awaited_by_is_set) {
+        assert(PySet_Check(_fut->fut_awaited_by));
+        int err = PySet_Discard(_fut->fut_awaited_by, thing);
+        if (err < 0 && PyErr_Occurred()) {
+            return -1;
+        } else {
+            return 0;
+        }
+    }
+    return 0;
+}
+
+static int
+future_awaited_by_discard(asyncio_state *state, PyObject *fut, PyObject *thing)
+{
+    if (!TaskOrFuture_Check(state, fut) || !TaskOrFuture_Check(state, thing)) {
+        // We only want to support native asyncio Futures.
+        // For further insight see the comment in the Python
+        // implementation of "future_add_to_awaited_by()".
+        return 0;
+    }
+
+    int result;
+    Py_BEGIN_CRITICAL_SECTION(fut);
```

Comment:
Same

Reply:
Check https://github.com/python/cpython/pull/124640#discussion_r1781868015

Reply:
What's your suggestion?

Reply:
> Also I believe nothing is to be gained for making this new feature thread safe while all the older code isn't.

Well exactly that: that this feature will be thread safe. Less thread unsafe code is strictly better than more thread unsafe code


> Currently whole future and task classes are unsafe to use in free-threaded builds. Last month iirc I fixed global access to be thread safe but not these classes, I haven't thought too much about this yet.

I don't agree with the argument that "the rest of the code is not thread safe so let's not fix this yet until we fix all at once". I think that if we have the opportunity to make the code correct from the get go we should do it and if someone wants to adapt the whole thing to a different approach later that can be done separately

Reply:
> Kumar, perhaps you should think about it. We use reviews to post concrete suggestions and finished thoughts.

My suggestion it to not make changes to free-threaded builds here in this PR, my unfinished thought is about what should be done for free-threaded builds separately which has nothing to do with this PR.

Reply:
I'm honestly mostly indifferent about this. @pablogsal your call.

Reply:
I don't feel very strongly so if you do, I can strip it out

Reply:
I removed them in https://github.com/python/cpython/pull/124640/commits/b9ecefb0477e12c70ead97bc628513cf12f07f43

Reply:
Alright, resolving this conversation!